diff --git a/include/clang/Basic/Attr.td b/include/clang/Basic/Attr.td
index d5ba722..e168bec 100644
--- a/include/clang/Basic/Attr.td
+++ b/include/clang/Basic/Attr.td
@@ -258,6 +258,7 @@ def TargetWindows : TargetArch<["x86", "x86_64", "arm", "thumb"]> {
 def TargetMicrosoftCXXABI : TargetArch<["x86", "x86_64", "arm", "thumb"]> {
   let CXXABIs = ["Microsoft"];
 }
+def TargetPRU : TargetArch<["pru"]>;
 
 class Attr {
   // The various ways in which an attribute can be spelled in source
diff --git a/lib/CodeGen/TargetInfo.cpp b/lib/CodeGen/TargetInfo.cpp
index 3d1ddef..0c0dcfb 100644
--- a/lib/CodeGen/TargetInfo.cpp
+++ b/lib/CodeGen/TargetInfo.cpp
@@ -6809,6 +6809,76 @@ SparcV9TargetCodeGenInfo::initDwarfEHRegSizeTable(CodeGen::CodeGenFunction &CGF,
 
 
 //===----------------------------------------------------------------------===//
+// PRU ABI Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+class PRUABIInfo : public DefaultABIInfo {
+public:
+  PRUABIInfo(CodeGenTypes &c) : DefaultABIInfo(c) {}
+
+  ABIArgInfo classify(QualType argty) const {
+    if (isAggregateTypeForABI(argty)) {
+      if (isEmptyRecord(getContext(), argty, true)) {
+        return ABIArgInfo::getIgnore();
+      }
+
+      // clpru's very convoluted struct passing rules:
+      // (pack, num_i32s_in_reg_block)  -- pack only if within same reg32
+      //     | struct_size <= 2 = (True, Just 1)
+      //     | struct_size <= 4 = (False, Just 1)
+      //     | struct_size == 8 = (False, Just 2)
+      //     | otherwise = (False, Nothing)  -- stored to stack
+      auto tysize = getContext().getTypeSize(argty);
+      /*
+      if (tysize == 8) {
+        return ABIArgInfo::getDirect(llvm::Type::getInt8Ty(getVMContext()));
+      } else if (tysize == 16) {
+        return ABIArgInfo::getDirect(llvm::Type::getInt16Ty(getVMContext()));
+      } else if (tysize == 24 || tysize == 32) {
+        return ABIArgInfo::getDirect(llvm::Type::getInt32Ty(getVMContext()));
+        //auto i8 = llvm::Type::getInt8Ty(getVMContext());
+        //return ABIArgInfo::getDirect(llvm::ArrayType::get(i8, 4));
+      } else if (tysize == 64) {
+        return ABIArgInfo::getDirect(llvm::Type::getInt64Ty(getVMContext()));
+      }*/
+      if (tysize <= 32 || tysize == 64) {
+        return ABIArgInfo::getDirect(nullptr, 0, nullptr, false);
+      } else {
+        return getNaturalAlignIndirect(argty);
+      }
+    }
+    if (const EnumType *EnumTy = argty->getAs<EnumType>()) {
+      argty = EnumTy->getDecl()->getIntegerType();
+    }
+    return (argty->isPromotableIntegerType() ? ABIArgInfo::getExtend()
+                                             : ABIArgInfo::getDirect());
+  }
+
+  void computeInfo(CGFunctionInfo &FI) const override {
+    if (!getCXXABI().classifyReturnType(FI)) {
+        QualType retty = FI.getReturnType();
+        if (retty->isVoidType()) {
+          FI.getReturnInfo() = ABIArgInfo::getIgnore();
+        } else {
+          FI.getReturnInfo() = classify(FI.getReturnType());
+        }
+    }
+    for (auto &I : FI.arguments()) {
+      I.info = classify(useFirstFieldIfTransparentUnion(I.type));
+    }
+  }
+};
+
+class PRUTargetCodeGenInfo : public TargetCodeGenInfo {
+public:
+  PRUTargetCodeGenInfo(CodeGenTypes &CGT)
+      : TargetCodeGenInfo(new PRUABIInfo(CGT)) {}
+};
+}
+
+//===----------------------------------------------------------------------===//
 // XCore ABI Implementation
 //===----------------------------------------------------------------------===//
 
@@ -7585,5 +7655,8 @@ const TargetCodeGenInfo &CodeGenModule::getTargetCodeGenInfo() {
     return *(TheTargetCodeGenInfo = new SparcV9TargetCodeGenInfo(Types));
   case llvm::Triple::xcore:
     return *(TheTargetCodeGenInfo = new XCoreTargetCodeGenInfo(Types));
+
+  case llvm::Triple::pru:
+    return *(TheTargetCodeGenInfo = new PRUTargetCodeGenInfo(Types));
   }
 }
diff --git a/lib/Basic/Targets.cpp b/lib/Basic/Targets.cpp
index af8aea0..e15c858 100644
--- a/lib/Basic/Targets.cpp
+++ b/lib/Basic/Targets.cpp
@@ -7426,6 +7426,67 @@ public:
   }
 };
 
+class PRUTargetInfo : public TargetInfo {
+public:
+  PRUTargetInfo(const llvm::Triple &Triple) : TargetInfo(Triple) {
+    BigEndian = false;
+    TLSSupported = false;
+    IntWidth = 32;
+    IntAlign = 8;
+    LongWidth = 32;
+    LongLongWidth = 64;
+    LongAlign = LongLongAlign = 8;
+    PointerWidth = 32;
+    PointerAlign = 8;
+    BoolWidth = 8;
+    BoolAlign = 8;
+    DefaultAlignForAttributeAligned = 8;
+    MinGlobalAlign = 8;
+    SuitableAlign = 8;
+    SizeType = UnsignedInt;
+    IntMaxType = SignedLongLong;
+    IntPtrType = SignedInt;
+    PtrDiffType = SignedInt;
+    SigAtomicType = SignedLong;
+    MaxTLSAlign = 8;
+    DataLayoutString = "e-S8-p:32:8:8-i32:8:8-i16:8:8-i8:8:8-a:8:8-n8:16:32";
+  }
+  void getTargetDefines(const LangOptions &Opts,
+                        MacroBuilder &Builder) const override {
+    Builder.defineMacro("PRU");
+    Builder.defineMacro("__PRU__");
+  }
+  ArrayRef<Builtin::Info> getTargetBuiltins() const override {
+    // FIXME: Implement.
+    return None;
+  }
+  bool hasFeature(StringRef Feature) const override { return Feature == "pru"; }
+  ArrayRef<TargetInfo::GCCRegAlias> getGCCRegAliases() const override {
+    // No aliases.
+    return None;
+  }
+  const char *getClobbers() const override {
+    // FIXME: Is this really right?
+    return "";
+  }
+  BuiltinVaListKind getBuiltinVaListKind() const override {
+    // FIXME: implement
+    return TargetInfo::CharPtrBuiltinVaList;
+  }
+  bool validateAsmConstraint(const char *&,
+                             TargetInfo::ConstraintInfo &) const override {
+    return false;
+  }
+  ArrayRef<const char *> getGCCRegNames() const override {
+    static const char *const GCCRegNames[] = {
+        "r0",  "r1",  "r2",  "r3",  "r4",  "r5",  "r6",  "r7",
+        "r8",  "r9",  "r10", "r11", "r12", "r13", "r14", "r15",
+        "r16", "r17", "r18", "r19", "r20", "r21", "r22", "r23",
+        "r24", "r25", "r26", "r27", "r28", "r29", "r30", "r31"};
+    return llvm::makeArrayRef(GCCRegNames);
+  }
+};
+
 class XCoreTargetInfo : public TargetInfo {
   static const Builtin::Info BuiltinInfo[];
 public:
@@ -7931,6 +7992,9 @@ static TargetInfo *AllocateTarget(const llvm::Triple &Triple) {
     if (!(Triple == llvm::Triple("wasm64-unknown-unknown")))
       return nullptr;
     return new WebAssemblyOSTargetInfo<WebAssembly64TargetInfo>(Triple);
+
+  case llvm::Triple::pru:
+    return new PRUTargetInfo(Triple);
   }
 }
